tools/indexer/main.go

```diff
@@ -0,0 +1,186 @@
+	dir := ""
+	if len(os.Args) > 1 {
+		// assume second arg is source dir
+		dir = os.Args[1]
+	} else {
+		// if no args specified assume we're running from the source dir
+		// and calculate the relative path to the services directory.
+		var err error
+		dir, err = filepath.Abs("../../services")
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	pkgs, err := getPackagesForIndexing(dir)
+	if err != nil {
+		panic(err)
+	}
+
+	indexedPkgs, err := getIndexedPackages()
+	if err != nil {
+		panic(err)
+	}
+
+	// for each package in pkgs, check if it's already been
+	// indexed.  if it hasn't been indexed then do so
+
+	for pkg := range pkgs {
+		if _, already := indexedPkgs[pkg]; already {
+			pkgs[pkg] = true
+			continue
+		}
+
+		fmt.Printf("indexing %s...", pkg)
+		resp, err := http.DefaultClient.Get(fmt.Sprintf("https://godoc.org/%s", pkg))
+		if err != nil {
+			panic(err)
+		}
+
+		resp.Body.Close()
+		if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+			// a 404 means that the package exists locally but not yet in github
+			fmt.Printf("completed with status '%s'\n", resp.Status)
+			pkgs[pkg] = true
+		} else {
+			fmt.Printf("FAILED with status '%s'\n", resp.Status)
+		}
+
+		// sleep a bit between indexing
+		time.Sleep(60 * time.Second)
+	}
+
+	// check if any packages failed to index
+	failed := false
+	for _, v := range pkgs {
+		if !v {
+			failed = true
+			break
+		}
+	}
+
+	if failed {
+		fmt.Println("not all packages were indexed")
+		os.Exit(1)
+	}
+	fmt.Println("successfully indexed all packages")
+}
+
+// returns the set of packages that have already been indexed
+func getIndexedPackages() (packageSet, error) {
+	// this URL will return the set of packages that have been indexed
+	resp, err := http.DefaultClient.Get("https://godoc.org/github.com/Azure/azure-sdk-for-go/services")
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("didn't receive 200 when looking for indexed packages, got %v: %s", resp.StatusCode, resp.Status)
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(body) < 1 {
+		return nil, errors.New("did't receive a response body when lookinig for indexed packages")
+	}
+
+	// scrape the response body to create the package list
+	pkgs := packageSet{}
+	bodyStr := string(body)
+	const lookFor = "github.com/Azure/azure-sdk-for-go/services/"
+	for {
+		// find the start of the package
+		index := strings.Index(bodyStr, lookFor)
+		if index < 0 {
+			break
+		}
+
+		// now find the terminal " character
+		terminal := strings.IndexRune(bodyStr[index+len(lookFor):], '"') + index + len(lookFor)
+		pkg := bodyStr[index:terminal]
+		pkgs[pkg] = true
+
+		// advance to the next
+		bodyStr = bodyStr[terminal:]
+	}
+	return pkgs, nil
+}
+
+// returns the set of packages, calculated from the repo, to be indexed
+func getPackagesForIndexing(dir string) (packageSet, error) {
```

Comment:
>getPackagesForIndexing [](start = 5, length = 22)

I was about to mention to use filepath.Glob but it does not support "**" :'(   [https://github.com/golang/go/issues/11862](https://github.com/golang/go/issues/11862)